zwave_js_server/util/node.py

```diff
@@ -115,12 +74,160 @@ async def async_set_config_parameter(
+            _validate_and_transform_new_value(zwave_value, partial_value)
+
+    cmd_response = await node.async_send_command(
+        "set_value",
+        valueId={
+            "commandClass": CommandClass.CONFIGURATION.value,
+            "property": property_,
+        },
+        value=new_value,
+    )
+    if cmd_response is False:
+        raise SetValueFailed(
+            "Unable to set value, refer to "
+            "https://zwave-js.github.io/node-zwave-js/#/api/node?id=setvalue for "
+            "possible reasons"
+        )
+
+    # If we received a response that is not false, the command was successful, otherwise
+    # we've queued it
+    return CommandStatus.ACCEPTED if cmd_response else CommandStatus.QUEUED
+
+
+async def async_set_config_parameter(
+    node: Node,
+    new_value: Union[int, str],
+    property_or_property_name: Union[int, str],
+    property_key: Optional[Union[int, str]] = None,
+) -> Tuple[ConfigurationValue, CommandStatus]:
+    """
+    Set a value for a config parameter on this node.
+
+    new_value and property_ can be provided as labels, so we need to resolve them to
+    the appropriate key
+    """
+    config_values = node.get_configuration_values()
+
+    # If a property name is provided, we have to search for the correct value since
+    # we can't use value ID
+    if isinstance(property_or_property_name, str):
+        try:
+            zwave_value = next(
+                config_value
+                for config_value in config_values.values()
+                if config_value.property_name == property_or_property_name
+            )
+        except StopIteration:
+            raise NotFoundError(
+                "Configuration parameter with parameter name "
+                f"{property_or_property_name} could not be found"
+            ) from None
+    else:
+        value_id = get_value_id(
+            node,
+            CommandClass.CONFIGURATION,
+            property_or_property_name,
+            endpoint=0,
+            property_key=property_key,
+        )
+
+        if value_id not in config_values:
+            raise NotFoundError(
+                f"Configuration parameter with value ID {value_id} could not be "
+                "found"
+            ) from None
+        zwave_value = config_values[value_id]
+
+    new_value = _validate_and_transform_new_value(zwave_value, new_value)
+
     # Finally attempt to set the value and return the Value object if successful
-    if await node.async_set_value(zwave_value, new_value) is False:
+    cmd_response = await node.async_set_value(zwave_value, new_value)
+    if cmd_response is False:
         raise SetValueFailed(
             "Unable to set value, refer to "
             "https://zwave-js.github.io/node-zwave-js/#/api/node?id=setvalue for "
             "possible reasons"
         )
 
-    return zwave_value
+    status = CommandStatus.ACCEPTED if cmd_response else CommandStatus.QUEUED
```

Comment:
I'm not sure this does what you expect (or I don't understand python enough).
When you call zwave-js's setValue API one of those things happens:
- `false` is returned if the API call cannot be done
- A `Promise` is returned if the API call can be done (and the message was queued)

The message stays in the queue until it is its turn to be sent. This can be almost immediately or worst case take hours until the device wakes up. Once the message was sent, again two things may happen (after a short while):
- The promise resolves to `true` if the command succeeded
- The promise rejects if the command did not succeed

For the calling code, this means `await`ing the call can result in three possible situations (for `cmd_response`):
- `false` (immediately): The call could not be done
- `true` (some time in the future - may be milliseconds, may be hours): The call succeeded
- Error (some time in the future - may be milliseconds, may be hours): The call failed

Reply:
We have two paths for sending a command: in one, we await the promise, in the other we don't. We determine which path to use based on the node status - if the node is asleep, we don't await the promise, if the node is any other status, we do.

When we await the promise, we expect true or false to be returned. If false, we raise an exception because something failed, if true, then in this PR we consider the `CommandStatus` to be `ACCEPTED`. When we don't await the promise, we assume that it's been queued. We have no way of verifying its future success/failure at that point. That logic is what has been implemented here.

I don't think I have seen the third situation you mentioned, the error scenario, but I don't think we account for it.

Reply:
I see.

> I don't think I have seen the third situation you mentioned, the error scenario, but I don't think we account for it.

Should be somewhat easy to test. Get a switchable plug, let it be interviewed, then pull it out of the socket before you try to change its configuration. The command will be handled immediately but fail due to a missing ACK.

Reply:
I just checked the server code, and we catch any errors that are thrown and return `success` as `False`, so I think we cover all three scenarios with this logic